Music widget waits for page to load before displaying.

michaelWagner opened the pull request:
fixes #987 and #989. 

Music widget doesn't load squished now. Waits for page to load before getting the dimensions of the parent div (card). Now the widget will always display properly for the size of the page it has been initially loaded for. 

There is still an issue that needs to be tackled seperately. Two cases: 
1. When the page is first loaded in the full two card style, the widget will look fine and all the sizes will be consistent. From there if the user resizes the page down to the one card style, the widget is still hidden and thus its height is incorrect when recalculating after the reisze event (I used the visibility property hidden instead of display:none and that seems to help, but there are still some problems with drastic resizing).
2. When the page is first loaded in the smaller one card style, the widget works well when sizing around within the one card window range. When some notes are added and the page is sized up to the full two card style and back down again a few times it will very intermittantly get squished. Still looking into why this is. 

I think having access to an event like isPanelVisible(PANEL_SUPPLEMENTAL) would be helpful so that we don't have to attach a listener to something as specific as a button click to know when the card is switched and the dimensions are ready to be calculated.

seanlip commented on extensions/interactions/MusicNotesInput/static/css/musicNotesInput.css:
```diff
@@ -8,6 +8,8 @@
   margin: 0 0 0 11.0784%;
   width: 87.34525%;
   height: 43%;
+  /*display: none; /* Hide valid note area until page loads */
```
This line can be removed entirely, right?

michaelWagner replied:
Ah yes! I thought the comment would be helpful toknow why it not visible initially but I guess it doesn't have to be there. Removed.

seanlip commented on extensions/interactions/MusicNotesInput/MusicNotesInput.js:
```diff
@@ -150,15 +150,49 @@ oppia.directive('oppiaInteractiveMusicNotesInput', [
+        // card switch event instead of the button click. Reinit needs to be
+        // triggered when page is in the smaller one card format or else
+        // dimensions for the staff will be incorrectly calculated.
+        $('.img-circle').on("click", function(e) {
```
Manipulating $('.img-circle') isn't great; it's really outside the scope of this interaction, and sets up a hidden dependency that will be hard to debug in the future (what happens if .img-circle is renamed or taken away, for example?). Better to pass events in from the container page.

Let's see first whether we need to do that in the scope of this commit, though. What additional regressions happen if you take the $('.img-circle') bits away?

michaelWagner replied:
Yes I agree. It really was the only way I could find to reliably fire the reinit when the page is in one card mode. If there is an event i could use to trigger it that would be ideal. Maybe the animation?  

When I take that $('.img-circle') block away the staff looks squished in the one card mode. I am open to suggestions on better approaches but I'll make some changes and see if I can get it better .

seanlip replied:
Have a look at conversationSkin.js -- there is a $scope.$broadcast('showInteraction'). This event is listened to in the InteractiveMap interaction; would it be useful for you? If so you can do something similar.

michaelWagner replied:
Thanks for the tip here. I added a listener to showInteraction and now I don't need to reinit on resize anymore because the parent div doesn't change anymore like it did when it was first designed. Works much better I think.

seanlip commented:
Btw, note that the Karma tests seem to be failing.

michaelWagner commented:
Thanks for the suggestion Sean. I think that is exactly what I'm looking
for. I'll look into why the tests failed as well.
On Wed, Nov 4, 2015 at 4:14 PM Sean Lip <notifications@github.com> wrote:

> Btw, note that the Karma tests seem to be failing.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/oppia/oppia/pull/1101#issuecomment-153886209>.
>

michaelWagner commented:
Ok I fixed the tests. Turns out that because the notes aren't being "physically" dropped on the staff in the karma test, they don't have a left position which means they aren't (nor do they have an opportunity to be) assigned a noteStart property either. I added it to the notes in the test and now they all pass. I'm going to dig into the showInteraction event tomorrow.

michaelWagner commented:
bash scripts/test.sh returned one error, though I don't think it was caused by me. 

```
5:55:13 FINISHED core.domain.user_jobs_test: 447.5 secs
----------------------------------------
----------------------------------------
Error 1
Traceback (most recent call last):
  File "/Users/michaelwagner/opensource/oppia/core/tests/gae_suite.py", line 119, in <module>
    main()
  File "/Users/michaelwagner/opensource/oppia/core/tests/gae_suite.py", line 101, in main
    suites = create_test_suites(parsed_args.test_target)
  File "/Users/michaelwagner/opensource/oppia/core/tests/gae_suite.py", line 71, in create_test_suites
    loader.loadTestsFromName(test_target)] if test_target
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/loader.py", line 100, in loadTestsFromName
    parent, obj = obj, getattr(obj, part)
AttributeError: 'module' object has no attribute 'classifier_services_test'


+------------------+
| SUMMARY OF TESTS |
+------------------+

SUCCESS   core.jobs_test: 29 tests (153.3 secs)
SUCCESS   core.controllers.admin_test: 5 tests (37.3 secs)
SUCCESS   core.controllers.base_test: 9 tests (64.5 secs)
SUCCESS   core.controllers.editor_test: 13 tests (172.4 secs)
SUCCESS   core.controllers.feedback_test: 10 tests (117.2 secs)
SUCCESS   core.controllers.galleries_test: 5 tests (67.4 secs)
SUCCESS   core.controllers.home_test: 9 tests (45.5 secs)
SUCCESS   core.controllers.moderator_test: 1 tests (9.7 secs)
SUCCESS   core.controllers.pages_test: 1 tests (2.8 secs)
SUCCESS   core.controllers.profile_test: 10 tests (32.5 secs)
SUCCESS   core.controllers.reader_test: 16 tests (169.8 secs)
SUCCESS   core.controllers.resources_test: 5 tests (51.9 secs)
= create_test_suites(parsed_args.test_target)
  File "/Users/michaelwagner/opensource/oppia/core/tests/gae_suite.py", line 71, in create_test_suites
    loader.loadTestsFromName(test_target)] if test_target
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/loader.py", line 100, in loadTestsFromName
    parent, obj = obj, getattr(obj, part)
AttributeError: 'module' object has no attribute 'classifier_services_test'
Traceback (most recent call last):
  File "scripts/backend_tests.py", line 322, in <module>
    main()
  File "scripts/backend_tests.py", line 282, in main
    test_count = int(tests_failed_regex_match.group(1))
AttributeError: 'NoneType' object has no attribute 'group'
```

seanlip commented on extensions/interactions/MusicNotesInput/MusicNotesInput.js:
```diff
@@ -150,15 +150,23 @@ oppia.directive('oppiaInteractiveMusicNotesInput', [
+        // displayed. The staffContainerElt and all subsequent measurements
+        // must be recalculated in order for the grid to work properly.
+        $scope.reinitStaff = function() {
+          // Prevent staff from reinitializing unneccesarily.
```
Does this comment still apply? It looks like the following line still executes regardless of any condition.

michaelWagner replied:
Oops I meant to remove that. Removed now.

seanlip commented on extensions/interactions/MusicNotesInput/static/css/musicNotesInput.css:
```diff
@@ -8,6 +8,7 @@
   margin: 0 0 0 11.0784%;
   width: 87.34525%;
   height: 43%;
+  visibility: hidden;
```
Ah -- I think you can keep the earlier comment, it is useful. The thing I was objecting to was the commented-out "display: none;" bit, which made it unclear whether that was supposed to be part of this CSS rule or not.

michaelWagner replied:
Okay, I see what you mean. I've added the comment back now.

seanlip commented:
Thanks, @michaelWagner -- looks good! Don't worry about the 1 backend test, this is a known issue (#1067, which should hopefully be fixed soon) and it looks like the Travis-CI checks work fine.

I made a couple of comments but after that it should be ready for merge.
